twine/package.py

```diff
@@ -72,29 +65,120 @@ def _safe_name(name: str) -> str:
     return re.sub("[^A-Za-z0-9.]+", "-", name)
 
 
-class CheckedDistribution(pkginfo.Distribution):
-    """A Distribution whose name and version are confirmed to be defined."""
+# Map ``metadata.RawMetadata`` fields to ``PackageMetadata`` fields.  Some
+# fields are renamed to match the names expected in the upload form.
+_RAW_TO_PACKAGE_METADATA = {
+    # Metadata 1.0 - PEP 241
+    "metadata_version": "metadata_version",
+    "name": "name",
+    "version": "version",
+    "platforms": "platform",  # Renamed
+    "summary": "summary",
+    "description": "description",
+    "keywords": "keywords",
+    "home_page": "home_page",
+    "author": "author",
+    "author_email": "author_email",
+    "license": "license",
+    # Metadata 1.1 - PEP 314
+    "supported_platforms": "supported_platform",  # Renamed
+    "download_url": "download_url",
+    "classifiers": "classifiers",
+    "requires": "requires",
+    "provides": "provides",
+    "obsoletes": "obsoletes",
+    # Metadata 1.2 - PEP 345
+    "maintainer": "maintainer",
+    "maintainer_email": "maintainer_email",
+    "requires_dist": "requires_dist",
+    "provides_dist": "provides_dist",
+    "obsoletes_dist": "obsoletes_dist",
+    "requires_python": "requires_python",
+    "requires_external": "requires_external",
+    "project_urls": "project_urls",
+    # Metadata 2.1 - PEP 566
+    "description_content_type": "description_content_type",
+    "provides_extra": "provides_extra",
+    # Metadata 2.2 - PEP 643
+    "dynamic": "dynamic",
+    # Metadata 2.4 - PEP 639
+    "license_expression": "license_expression",
+    "license_files": "license_file",  # Renamed
+}
+
+
+class PackageMetadata(TypedDict, total=False):
 
+    # Metadata 1.0 - PEP 241
+    metadata_version: str
     name: str
     version: str
+    platform: List[str]
+    summary: str
+    description: str
+    keywords: List[str]
+    home_page: str
+    author: str
+    author_email: str
+    license: str
+
+    # Metadata 1.1 - PEP 314
+    supported_platform: List[str]
+    download_url: str
+    classifiers: List[str]
+    requires: List[str]
+    provides: List[str]
+    obsoletes: List[str]
+
+    # Metadata 1.2 - PEP 345
+    maintainer: str
+    maintainer_email: str
+    requires_dist: List[str]
+    provides_dist: List[str]
+    obsoletes_dist: List[str]
+    requires_python: str
+    requires_external: List[str]
+    project_urls: Dict[str, str]
+
+    # Metadata 2.1 - PEP 566
+    description_content_type: str
+    provides_extra: List[str]
+
+    # Metadata 2.2 - PEP 643
+    dynamic: List[str]
+
+    # Metadata 2.4 - PEP 639
+    license_expression: str
+    license_file: List[str]
+
+    # Additional metadata
+    comment: Optional[str]
+    pyversion: str
+    filetype: str
+    gpg_signature: Tuple[str, bytes]
+    attestations: str
+    md5_digest: str
+    sha256_digest: Optional[str]
+    blake2_256_digest: str
```

Comment:
These aren't actually metadata in the way the rest of these fields are. These are internal-ish data about the package file in question. In other words, there's a real and valuable distinction between "I parsed this from the package metadata" and "I generated this from the package file to be uploaded and/or its metadata". Blurring those lines may be convenient-ish, but long term will be confusing for new contributors, maintainers, etc. I don't believe these should live here. As I said, we can use packaging's typed dict as a collaborator here, but I won't agree to inheriting from it or doing something like this

Reply:
Thanks @sigmavirus24 for the review. However, I think there is a misunderstanding.

The type definition you are commenting on is defining the return type of the `Package.metadata_dictionary()`: https://github.com/pypa/twine/blob/1703ae7316f73358abcc01c2cf0df6b922094a3f/twine/package.py#L178-L182

As you can see, the goal of the function is indeed to merge multiple sources of "metadata". This is existing code. The only thing that my patch does it to type the returned value as a TypedDict. I thought that using a TypedDict made things clearer and resolved some ambiguities regarding the data types stored in the dictionary. If adding typing is more confusing than helpful, I can remove it. I agree that the naming of the method is not the most clear: what is returned is the package submission data, not the package metadata, but, again, that is the existing name of the method. I named the TypedDict to match the method name. I can change that to something else, if we can find a more fitting name.

> As I said, we can use packaging's typed dict as a collaborator here, but I won't agree to inheriting from it or doing something like this

Things are already structured like this: the `Package` class has a `metadata` member that holds the parsed metadata as returned by `packaging` and the additional fields required for submission: https://github.com/pypa/twine/blob/1703ae7316f73358abcc01c2cf0df6b922094a3f/twine/package.py#L82-L96
or in the proposed patch:
```python
class PackageFile:
    def __init__(
        self,
        filename: str,
        comment: Optional[str],
        metadata: metadata.RawMetadata,
        python_version: str,
        filetype: str,
    ) -> None:
        self.filename = filename
        self.basefilename = os.path.basename(filename)
        self.comment = comment
        self.metadata = metadata
        self.python_version = python_version
        self.filetype = filetype
```
However, at some point the different sources of data for the submission form need to be combined into a single data structure. The existing code uses a dictionary for that.

I hope that this clarification resolves the issue you have with the code structure. If not, I may be misunderstanding your comment. In this case I would like to point out that all the proposed patch does is to change the source of the metadata values. The combining into a single dictionary of all the value required for submission is existing code. If you prefer to structure this code differently, maybe you can change the existing code to a structure you like more, and I'll rebase my changes on top.

Reply:
Ah that does clarify things. Sorry for the confusion. I saw our new maintainers approval and could only review from my phone which makes it harder to follow some things. Sorry for the confusion.

Reply:
> Sorry for the confusion.

No problem. Thank you for the review and for merging!